packages/core/src/auth/email.ts

```diff
@@ -15,6 +15,10 @@ export class EmailAuthServer implements AuthServer {
         authenticator.state = { email };
         if (authenticator.status !== AuthenticatorStatus.Active) {
             authenticator.state.activationCode = await this._generateCode();
+            // Override token if running e2e tests
+            if (process.env.PL_E2E_TESTS && process.env.PL_E2E_TESTS === "true" && email === "user@example.com") {
+                authenticator.state.activationCode = "000000";
+            }
```

Comment:
I know this is probably just a stop-gap solution, but the spirit of a true e2e test, it would be better to find a way to simulate the code being delivered to the client instead bypassing the mechanism. I would also like to avoid having testing-related code inside the components to be tested.

In the old integration tests, I was using the [`StubMessenger`](https://github.com/padloc/padloc/blob/v4/packages/core/src/messenger.ts#L85) email backend to store the code in memory and then retrieve it later: https://github.com/padloc/padloc/blob/master/packages/core/src/spec/app.ts#L53
The "stub" backend isn't currently supported by the new config scheme but it could easily be added here: https://github.com/padloc/padloc/blob/v4/packages/server/src/init.ts#L116

Of course this only works if the server memory is accessible by the test code, which I don't know whether that's the case here. Alternatively, we could use the [ConsoleMessenger](https://github.com/padloc/padloc/blob/v4/packages/core/src/messenger.ts#L105) backend and parse the code from stdout (again don't know if that's feasible here).

You can use the `PL_EMAIL_BACKEND` variable to set the email driver.

Reply:
I understand what you mean and generally agree, and I'd add it's better to have as little differences in the code as possible.

This still runs everything as "in reality", except we simply force the code to be known.

I very much prefer something like this over mocking a whole piece of technology or using a different one. Generally, the less differences between a test and "reality", the better.

Does that make sense?

Reply:
I hear what you're saying, but I don't agree with your conclusion. I agree that mocking is often overused, but the fact is we need **some** kind of mocking no matter what (unless we want to go as far as actually fetching and parsing emails from a real mailbox). In this case you're mocking the generation of activation code, whereas I would rather mock the delivery mechanism.

"You have to die one kind of death", as we say in Germany, and in this case I do think that mocking the email backend is the less painful one. Despite the shamefully bad test coverage, Padloc was actually designed to make this kind of mocking fairly trivial (in this case it is really just a configuration change), and I think we should use this to our advantage 😊

Reply:
Thanks @MaKleSoft ! I understand your point, and while I don't agree, I'd rather move on with the useful part here (the testing) instead of spending more time in the implementation _for now_, unless you disagree.

From my understanding, mocking the delivery mechanism + reading from it will not be trivial _at all_ (and I'll probably need to implement some new endpoint to do so, because the browser can't access the server's memory otherwise), and it'll likely take a few days to do, whereas right now _that part is working_, just in a different way than your preferred, and it isn't exposing any security vulnerability since it's checked from an environment variable + hardcoded test email.

I'm happy to implement your preferred mocking method later, after the test itself is working and verified, if you agree?

Reply:
Oh I totally agree that we can leave this as-is for now, especially if my suggested method turns out to come with a significant overhead. Like I said, I don't even know if my approach is feasible at all in this case, since it would require either direct access to the server or some way to read stdout. Let's revisit this discussion once the PR is ready for review 👍